src/jltypes.c

```diff
@@ -400,69 +398,190 @@ static jl_value_t *intersect_union(jl_uniontype_t *a, jl_value_t *b,
     return tu;
 }
 
-// if returns with *bot!=0, then intersection is Union{}
-static size_t tuple_intersect_size(jl_svec_t *a, jl_svec_t *b, int *bot)
+/*
+Simplification of varargs tuple types:
+ JL_TUPLE_FIXED: tuples of known length (e.g., JL_VARARG_NONE or JL_VARARG_INT)
+ JL_TUPLE_VAR:   tuples of unknown length (e.g., JL_VARARG_BOUND or JL_VARARG_UNBOUND)
+
+In some cases, JL_VARARG_BOUND tuples get described as JL_TUPLE_FIXED,
+if the constraints on length are already known.
+
+lenr = "representation length" (the number of parameters)
+lenf = "full length" (including the Vararg length, if known)
+
+In general, lenf >= lenr-1. The lower bound is achieved only for a Vararg of length 0.
+*/
+typedef enum {
+    JL_TUPLE_FIXED = 0,
+    JL_TUPLE_VAR   = 1
+} JL_TUPLE_LENKIND;
```

Comment:
Was just going to fix this, and noticed that there's also `JL_VARARG_KIND`, which I assume should have the same capitalization as `JL_TUPLE_LENKIND`? I happen to have a function called `jl_vararg_kind`. I could come up with another name, of course. However, I'm noting that enums defined in julia.h seem to be uppercase (even the "typename"), whereas enums defined outside julia.h seem to be lowercase. Is this historical or intentional? (I can clean other stuff up along the way, if you like.)